### DISCUSSION TITLE
Potential scenarios and solutions with sharing data between files

### DISCUSSION THREAD

**Discussion by gavinmcfarland** [fignite/helpers#3] [Ideas]
Below I'd like to discuss and explore the different scenarios, problems and solutions available to sharing data between files.

These are the two things I can think of that affect this:

1. Whether it is the same user, or multiple users
2. Whether the file has been published or not

Below are some solutions to being able to access data from a different file:

r  | Solution | Benefits | Disadvantages
-- | -- | -- | --
A | When a user runs your plugin save an object/array to clientStorage. When they open the plugin on another file the plugin now has access to things/components in that file. | User doesn’t need to provide any information | Only works between same users. Doesn’t work if they only have view access to that file. Doesn't work across devices. Will be lost if the user clears their cache, or logs out.
B | Store data on a component and publish the component. When the plugin is run in another file, so long as an instance of that component has been created by the user, you can now use the plugin to get the data from the component. | Convenient way of sharing data between files that doesn’t require credentials. Might work in most cases because user may have already created an instance of component | Requires the user to create an instance of the component. Have to teach and explain this to the user
C | Ask the user to provide a fileKey or full url to the file and generate an access token. Using these two things you can now get the entire contents of the file to import what you want. | Only receiving user needs to add details of file and generate an access token. Possible that other plugins could re-use this access token | Could other plugins use this access token nefariously?
D | Ask the user to store information on a server. Then from the other file when the plugin is run ask the user to provide the url (and maybe access token) the server (like json.bin) and get the information you need. | Easiest to understand? | Both sending and receiving user need to provide details. Requires using or setting up an external server
E | Ask user to provide key of component they want to import. Could be made easier by creating a data component which is kept in sending file and plugin giving user key to that component. Then from another file plugin can ask for this key. | Avoids needing an external serverAvoids needing an access token | Component is at risk of being deletedUser must shared key with others
F | Collect data from each user which is stored in pluginData of each file the plugin is used on. Each new user will then already have access to components from files they’ve never visited. | Requires very little effort from user. Increases chances that plugin will have access to component that’s needed. | Could add unnecessary size to each file visited by user. Would need to sync this potentially, could get messy?

**Comment by LukeFinch** [comment]
Line A also has the issue of being lost if a user clears their cache. And won't persist cross-device.

**Comment by gavinmcfarland** [comment]
In my most recent work I've found this setup works well.

1. Store file data on the root
2. Create an empty array for remote files on the root
3. Push file data to an array of files in clientStorage
4. In remote file sync new file data from clientStorage to array on the root of the remote file

I'd like to create some helpers with this, but it's quite complex in reality so it's hard to know how to abstract it.
